### ISSUE TITLE
Clusterio not closing file descriptors leading to EMFILE

### ISSUE THREAD

**Issue by mlyszczek** [clusterio/factorioClusterio#306]
It appears cluesterio client opens file but never closes it.

```
# lsof -i -n -P  | grep 28290 | wc -l
2715
```

28290 is a pid of clusterio-client. This leads to clusterio crash after few hours due to EMFILE error.

```
internal/fs/utils.js:259
    throw err;
    ^

Error: EMFILE: too many open files, open './instances/gentoo-pl/script-output/researchSync.txt'
    at Object.openSync (fs.js:461:3)
    at Object.readFileSync (fs.js:364:35)
    at Timeout._onTimeout (/home/factorio/clusterio-client/client.js:654:52)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7) {
  errno: -24,
  syscall: 'open',
  code: 'EMFILE',
  path: './instances/gentoo-pl/script-output/researchSync.txt'
}
```

```
$ git status
HEAD detached at 1.2.4
nothing to commit, working tree clean
```

```
$ node -v
v14.4.0
```

I happens on Gentoo, I followed install guide from readme.md and changed nothing in source code. I know virtually nothing about  nodejs nor js, so I can't really debug this myself. If you need to me share anything else, just say so.

**Comment by Hornwitser** [comment]
Which files is it creating all the handles to?

**Comment by mlyszczek** [comment]
Further investigation shows these file descriptors are in fact TCP connections to clusterio master that were not closed.

```
# ls -l /proc/28290/fd/ | head -n10
total 0
lrwx------ 1 factorio factorio 64 sie 25 23:08 0 -> /dev/null
l-wx------ 1 factorio factorio 64 sie 25 23:08 1 -> /var/log/clusterio-slave.log
lr-x------ 1 factorio factorio 64 sie 25 23:08 10 -> pipe:[644177013]
lrwx------ 1 factorio factorio 64 sie 25 23:08 100 -> socket:[644330919]
lrwx------ 1 factorio factorio 64 sie 25 23:08 1000 -> socket:[646126087]
lrwx------ 1 factorio factorio 64 sie 25 23:08 1001 -> socket:[646128763]
lrwx------ 1 factorio factorio 64 sie 25 23:08 1002 -> socket:[646128137]
lrwx------ 1 factorio factorio 64 sie 25 23:08 1003 -> socket:[646131880]
lrwx------ 1 factorio factorio 64 sie 25 23:08 1004 -> socket:[646132259]
```

```
# lsof -i -a -p 28290 | head -n10
COMMAND   PID     USER   FD   TYPE    DEVICE SIZE/OFF NODE NAME
node    28290 factorio   19u  inet 644179610      0t0  TCP localhost:52684->localhost:5225 (ESTABLISHED)
node    28290 factorio   21u  inet 644178829      0t0  TCP kurwik.kurwinet.pl:51464->kurwik.kurwinet.pl:8082 (ESTABLISHED)
node    28290 factorio   23u  inet 644178817      0t0  TCP kurwik.kurwinet.pl:51450->kurwik.kurwinet.pl:8082 (ESTABLISHED)
node    28290 factorio   25u  inet 644182144      0t0  TCP kurwik.kurwinet.pl:51526->kurwik.kurwinet.pl:8082 (ESTABLISHED)
node    28290 factorio   26u  inet 644182671      0t0  TCP kurwik.kurwinet.pl:51606->kurwik.kurwinet.pl:8082 (ESTABLISHED)
node    28290 factorio   27u  inet 644186350      0t0  TCP kurwik.kurwinet.pl:51688->kurwik.kurwinet.pl:8082 (ESTABLISHED)
node    28290 factorio   28u  inet 644187338      0t0  TCP kurwik.kurwinet.pl:51754->kurwik.kurwinet.pl:8082 (ESTABLISHED)
node    28290 factorio   29u  inet 644190388      0t0  TCP kurwik.kurwinet.pl:51836->kurwik.kurwinet.pl:8082 (ESTABLISHED)
node    28290 factorio   30u  inet 644189922      0t0  TCP kurwik.kurwinet.pl:51888->kurwik.kurwinet.pl:8082 (ESTABLISHED)
```

```
# netstat -plant | grep 28290 | head -n10 
tcp        0      0 10.1.1.1:46426          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie 
tcp        0      0 10.1.1.1:46998          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie 
tcp        0      0 10.1.1.1:53440          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie 
tcp        0      0 10.1.1.1:48584          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie 
tcp        0      0 10.1.1.1:44514          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie 
tcp        0      0 10.1.1.1:41108          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie 
tcp        0      0 10.1.1.1:35322          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie 
tcp        0      0 10.1.1.1:33812          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie 
tcp        0      0 10.1.1.1:44232          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie 
tcp        0      0 10.1.1.1:42016          10.1.1.1:8082           ESTABLISHED 28290/clusterioClie
```

**Comment by mlyszczek** [comment]
It appears that new connection is created once every 10 seconds. Could be problem with
```
setInterval(B=> socket.emit("heartbeat"), 10000);
```
in instanceManagement()?

**Comment by psihius** [comment]
Hmm, well, heartbeat is expected, but that it creates new connections is weird.

**Comment by psihius** [comment]
Yeah, this is something socketio does on its own and at a glance does not look like there is something wrong with the code.
Though we have run some major clusters with like 40-50 servers with no downtime in weeks and we never encountered this issue.
So my question would be, can it be that some network settings in the OS got tweaked and it causes this issue? Like huge socket closing timeout or socket re-use is not configured?  (I used to play with Gentoo too, I know how deep the rabbit hole with tweaking it goes :D).

**Comment by mlyszczek** [comment]
No, these options are default. But I've narrowed down the problem. It's not heartbeat, it's hashMods() function, something there establishes connection each time function getID() is called.

**Comment by mlyszczek** [comment]
Hi,

I found a solution for this and created pull request #307 

Basically needle module waited forever to get response for /api/getID call. This is because in `node_modules/needle/lib/needle.js`
in defaults.response_timeout there is '0' which is not overwritten, and when set_timeout() function is called there is check `if (milisecs <= 0) return;` so when timeout is set to 0, timeout is not installed.

No idea how this does not crash other machines and why it appears like I am the only one affected by it.

**Comment by Hornwitser** [comment]
We should probably set the response timeout to something sensible.

**Comment by Hornwitser** [comment]
Fixed in #307
